rust_src/src/multibyte.rs

```diff
@@ -42,32 +43,285 @@ use crate::{
     hashtable::LispHashTableRef,
     lisp::{ExternalPtr, LispObject, LispStructuralEqual},
     obarray::LispObarrayRef,
-    remacs_sys::Qstringp,
     remacs_sys::{
         char_bits, equal_kind, EmacsDouble, EmacsInt, Lisp_Interval, Lisp_String, Lisp_Type,
     },
     remacs_sys::{compare_string_intervals, empty_unibyte_string, lisp_string_width},
+    remacs_sys::{Qcharacterp, Qstringp},
     symbols::LispSymbolRef,
 };
 
 pub type LispStringRef = ExternalPtr<Lisp_String>;
 
-// cannot use `char`, it takes values out of its range
-pub type Codepoint = u32;
-
 /// Maximum character code
-pub const MAX_CHAR: Codepoint = (1 << char_bits::CHARACTERBITS as usize) - 1;
+pub const MAX_CHAR: u32 = (1 << char_bits::CHARACTERBITS as usize) - 1;
+
+// Maximum character codes for several encoded lengths
+/// Maximum value for a single byte codepoint
+pub const MAX_1_BYTE_CHAR: u32 = 0x7F;
+
+/// Minimum value for a two byte codepoint
+pub const MIN_2_BYTE_CHAR: u32 = 0x80;
+/// Minimum value for a two byte codepoint
+pub const MAX_2_BYTE_CHAR: u32 = 0x7FF;
 
-/// Maximum character codes for several encoded lengths
-pub const MAX_1_BYTE_CHAR: Codepoint = 0x7F;
-pub const MAX_2_BYTE_CHAR: Codepoint = 0x7FF;
-pub const MAX_3_BYTE_CHAR: Codepoint = 0xFFFF;
-pub const MAX_4_BYTE_CHAR: Codepoint = 0x1F_FFFF;
-pub const MAX_5_BYTE_CHAR: Codepoint = 0x3F_FF7F;
+/// Minimum value for a three byte codepoint
+pub const MIN_3_BYTE_CHAR: u32 = 0x800;
+/// Maximum value for a three byte codepoint
+pub const MAX_3_BYTE_CHAR: u32 = 0xFFFF;
+
+/// Minimum value for a four byte codepoint
+pub const MIN_4_BYTE_CHAR: u32 = 0x10000;
+/// Maximum value for a four byte codepoint
+pub const MAX_4_BYTE_CHAR: u32 = 0x1F_FFFF;
+
+/// Minimum value for a five byte codepoint
+pub const MIN_5_BYTE_CHAR: u32 = 0x20_0000;
+/// Maximum value for a five byte codepoint
+pub const MAX_5_BYTE_CHAR: u32 = 0x3F_FF7F;
+
+pub const MIN_BYTE8_CHAR: u32 = 0x3F_FF80;
+pub const MAX_BYTE8_CHAR: u32 = 0x3F_FFFF;
 
 /// Maximum length of a single encoded codepoint
 pub const MAX_MULTIBYTE_LENGTH: usize = 5;
 
+const BYTE8_OFFSET: u32 = 0x3F_FF00;
+
+// cannot use `char`, it takes values out of its range
+#[repr(transparent)]
+#[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Debug)]
+pub struct Codepoint(u32);
+
+impl Codepoint {
+    // Equivalent to BYTE8_TO_CHAR
+    /// Create a codepoint from a raw byte. If non-ascii, byte8 encode it.
+    pub fn from_raw(byte: u8) -> Codepoint {
+        match Codepoint::from(byte) {
+            cp if cp.is_ascii() => cp,
+            cp => Codepoint::from(cp.0 + 0x3F_FF00),
```

Comment:
Should use the constant `BYTE8_OFFSET` which you added

Reply:
Fixed